AI/BattleAI/BattleAI.cpp

```diff
@@ -554,12 +558,127 @@ void CBattleAI::print(const std::string &text) const
 
 boost::optional<BattleAction> CBattleAI::considerFleeingOrSurrendering()
 {
-	if(cb->battleCanSurrender(playerID))
+	LOGL("Retreat sounds like fun. Let's see...");
+	using ValueMap = PossibleSpellcast::ValueMap;
+	auto isItOurLastTurn = [&](ValueMap & values, const std::vector<battle::Units> & queue, HypotheticBattle * state) -> bool
 	{
+		bool firstRound = true;
+		bool ourFirstTurn = true;
+		bool enemyHadTurn = false;
+
+		for(auto & round : queue)
+		{
+			if (!firstRound)
+			{
+				state->nextRound(0);//todo: set actual value?
+				firstRound = false;
+			}
+			for(auto unit : round)
+			{
+				LOGL("---------------> consider turn of " + unit->unitType()->namePl);
+				if (!state->getForUpdate(unit->unitId())->alive())

```

Comment:
Just `(!unit->alive())` should be anough

Reply:
well, it's not. If unit is alive now, but is killed in the predicted scenario, `unit->alive()` remains true in contrast to what is done here. But I'll make sure, thanks.

Reply:
I see now, you need to call battleGetTurnOrder on HypotheticBattle  but not `Callback`

Reply:
I double checked this thing. There has to be `state->getForUpdate(..)->alive`. Otherwise, it can return alive even if we predict the unit will die.
it also has to be fixed in spellcasting prediction